docs/style/CODING_STYLE_GUIDE.md

```diff
@@ -0,0 +1,162 @@
+# Coding Style Guide
+
+_Revision 6_ _2024-10-28_
+
+This guide provides a small set of code guidelines used in the SDK. This guide
+reflects the currently accepted style practices for the SDK and is subject to
+change.
+
+The SDK was seeded from multiple different projects and contains contributions
+from many different companies and the SDK code therefore uses several different
+coding styles throughout the code base. Stylistically, code should attempt to
+conform to the dominant style of the code being modified, while also adhering to
+the guidelines below.
+
+## Language standard
+
+Code in the SDK conforms to the following standards. Changes to the C++ standard
+happen infrequently and there has been only one change from C++14 to C++17 over
+the course of the SDK lifetime. Changes to the Python version happen more
+frequently.
+
+| Language | Version |
+| -------- | ------- |
+| C++      | C++17   |
+| Python   | 3.10    |
+
+Product-specific software may elect to use later standards in their own work.
+
+## Coding Guidelines
+
+### Common
+
+#### When in Rome
+
+The most important convention and practice in the Matter SDK repo is "_When in
+Rome..._", per the quote below.
+
+[quote, St. Ambrose]
+
+---
+
+If you should be in Rome, live in the Roman manner; if you should be elsewhere,
+live as they do there.
+
+---
+
+Your extensions or fixes to existing code should match the prevailing of the
+original code.
+
+If you find the conventions so foreign or otherwise confusing, it may be best to
+let whoever owns the file make the necessary changes or seek the counsel of
+others in the group to find out what the right thing to do is. Never just start
+changing code wholesale for personal reasons without consulting others first.
+
+#### Commenting Out or Disabling Code
+
+Unused code shall not be disabled by commenting it out with C- or C++-style
+comments or with preprocessor `#if 0 ... #endif` semantics.
+
+#### Auto-formatters
+
+We use the following auto-formatters on code:
+
+| Language   | Formatter          | Style File                                                                                                  |
+| ---------- | ------------------ | ----------------------------------------------------------------------------------------------------------- |
+| C++        | clang-format       | [.clang-format](https://github.com/project-chip/connectedhomeip/blob/master/.clang-format)                  |
+| ObjectiveC | clang-format       | [.clang-format](https://github.com/project-chip/connectedhomeip/blob/master/.clang-format)                  |
+| java       | google-java-format | N/A                                                                                                         |
+| Python     | pep8, isort        | [.restyled.yaml](https://github.com/project-chip/connectedhomeip/blob/master/.restyled.yaml) (command line) |
```

Comment:
.restyled.yaml is not a style file for python, it's a configuration for restyled. However I can see that it has command line options for autopep8.... which is not perfect because autopep8 run locally will not see these options... How about moving them to one of these: https://github.com/hhatto/autopep8?tab=readme-ov-file#id8 Also, we have `ruff` rules: https://github.com/project-chip/connectedhomeip/blob/master/ruff.toml

config for isort: https://github.com/project-chip/connectedhomeip/blob/master/.isort.cfg

Reply:
Added config for isort. Going to defer the move of the autopep stuff to avoid mixing documentation and CI changes in the same PR.

So we RUN ruff in the CI? I'm not sure I've ever failed it and I'd DEFINITELY done some weird stuff that should have failed it.

Reply:
> So we RUN ruff in the CI?

https://github.com/project-chip/connectedhomeip/blob/master/.github/workflows/lint.yml#L307

> I've ever failed it and I'd DEFINITELY done some weird stuff that should have failed it.

It has lots of warning disabled. Most of these warning are style related, so the CI will not fail on "style" issues, but only on potential code bugs. The style, IMHO, should be handled by restyled, which provides a patch for a fixup. I was never configured though, because as far as I know there are some issues with line length automatic reflow. But maybe it has changed now...

Reply:
oh, the linter. Cool. OK, I'll add that - thanks